mirror-action: Adding User Guide (#276)
foundriesio/docs: source/user-guide/mirror-action/mirror-action.rst

.. _ug-mirror-action:

> Mirror Action

Reviewer: I have mixed feelings about this title. On one hand, this section is focusing GitHub, on the other hand, the reader can use the same instructions with few changes with several other git hosting (as you mention)

I'm not sure about this suggestion, but maybe it would be more informative to use a title such as:
Configuring Automatic Git Mirroring

I'm not sure, because, although the title is not informative, the very first line of the description is very informative, so I'm not sure

Author: mirror-action is not something exclusive for Github. The name itself doesn't come from Github, however, I accept your suggestion and agree that Configuring Automatic Git Mirroring sounds better.

Reviewer: "Action" is the word GitHub uses for the automatic processes, GitLab use something different, Jobs or Pipeline, I don't remember

Author: Got it, so makes sense. I already changed it.

=============

This section shows how to configure a GitHub_ Action for mirroring your commits 
to the FoundriesFactory repository.

> It is very useful when you want to use an external private or public git repository 

Reviewer 2: ```suggestion
This is very useful when you want to use an external private or public git repository,
```

such as GitHub_.

> This section could be adapted for other git repository hosting services.

Reviewer 2: ```suggestion
This section can be adapted for other git repository hosting services.
```

Creating Token
--------------

To allow GitHub_ to access your FoundriesFactory repository, you need to create a token.

Go to `Tokens <https://app.foundries.io/settings/tokens/>`_ and create a new **Api Token** by clicking on 
:guilabel:`+ New Token`.

Complete with a **Description** and the **Expiration date** and select :guilabel:`next`.

> For GitHub_, check the :guilabel:`Use for source code access` box and 
> select your **Factory**. Remember that you can revoke this access and set up a new credential later once you are
> familiar with the :ref:`ref-api-access`.

Reviewer 2: I don't see the need for this here
```suggestion
For GitHub_, check the :guilabel:`Use for source code access` box and 
select your **Factory**.
```

.. figure:: /_static/userguide/mirror-action/mirror-action.png
   :width: 500
   :align: center

   Token for source code access

Convert the token value to a base64 string:

.. prompt:: bash host:~$, auto

    host:~$ echo -n <FIO_TOKEN> | base64 -w0

**Example Output**:

.. prompt:: text

     host:~$ echo -n SQMD1Gx860mPI6jZFlLJLwaCXT5CqAaQi6nEfIfH | base64 -w0
     U1FNRDFHeDg2MG1QSTZqWkZsTEpMd2FDWFQ1Q3FBYVFpNm5FZklmSA==

Save the output of this command to your copy buffer. This is your ``<BASE64_FIO_TOKEN>`` value.

.. note::

    Your value should end with ``==`` and the output won't have a carriage return added.

Configuring GitHub Repository
-----------------------------

Go to GitHub_ and find the repository you want to mirror.

> Click on :guilabel:`Settings`

Reviewer 2: ```suggestion
Click on :guilabel:`Settings`:
```

.. figure:: /_static/userguide/mirror-action/mirror-action-github-setting.png
   :width: 900
   :align: center

   GitHub Settings

Click on :guilabel:`Secrets` and create a new **secret** by clicking on 
:guilabel:`New repository secret`.

.. figure:: /_static/userguide/mirror-action/mirror-action-github-secrets.png
   :width: 900
   :align: center

   GitHub Secrets

> Complete the **Name** with ``GIT_ACCESS_TOKEN`` and on **Value** paste the ``<BASE64_FIO_TOKEN>``.

Reviewer 2: ```suggestion
Complete the **Name** with ``GIT_ACCESS_TOKEN`` and on **Value** paste the ``<BASE64_FIO_TOKEN>``
provided above.
```

Reviewer: Is GIT_ACCESS_TOKEN only an example of name or it must have this name to actually work?

Finally, click on :guilabel:`Add secret`.

.. figure:: /_static/userguide/mirror-action/mirror-action-github-new-secret.png
   :width: 900
   :align: center

   GitHub New Secret

Creating Mirror Action
----------------------

> Update the clone command with your repository URL and enter its directory:
> 
> .. prompt:: bash host:~$
> 
>     git clone https://github.com/<host>/<repo_name>
>     cd <repo_name>

Reviewer 2: I think it's better to clone the repository after the factory-config explanation.
```suggestion
```

> The FoundriesFactory CI just triggers builds for configured branches. This is 
> configured in the ``ci-script.git`` repository in the ``factory-config.yml`` file.

Reviewer 2: ```suggestion
The FoundriesFactory CI only triggers builds for configured branches. This is 
configured in the ``ci-scripts.git`` repository in the ``factory-config.yml`` file.
```

> Your ``factory-config.yml`` could be inspected by updating the following URL with your ``<FACTORY-NAME>``:

Reviewer 2: ```suggestion
Your ``factory-config.yml`` can be inspected by updating the following URL with your ``<FACTORY-NAME>``:
```

- https://source.foundries.io/factories/<FACTORY-NAME>/ci-scripts.git/tree/factory-config.yml

> Under ``lmp:`` and ``containers:`` the group ``tagging:`` will contain the configured branches on ``refs/heads/<branch>:``.

Reviewer 2: ```suggestion
Under ``lmp:`` and ``containers:`` the group ``tagging:`` shows the configured branches on ``refs/heads/<branch>:``.
```

**Example**:

.. prompt:: text

    lmp:
      tagging:
        refs/heads/master:
          - tag: master
        refs/heads/devel:
          - tag: devel
    ...
    containers:
      tagging:
        refs/heads/master:
          - tag: master
        refs/heads/devel:
          - tag: devel
    ...

> Based on the example, FoundriesFactory CI is configured to build pushes from 
> ``master`` and ``devel``. The following commands, will guide you to mirror the 
> ``master`` branch.

Reviewer 2: ```suggestion
Based on this example, FoundriesFactory CI is configured to build pushes from 
``master`` and ``devel``. The following commands will guide you to mirror the 
``master`` branch.

Clone your GitHub repository and enter its directory:

.. prompt:: bash host:~$

    git clone https://github.com/<host>/<repo_name>
    cd <repo_name>

.. note::

    Make sure to update the clone command with your repository URL.
    
```

Reviewer: the CI does not build pushes, it builds source code. CI builds are triggered by new commits (which can be understood as pushes).

The following commands happens in the present, not in the future (the commands are there already)

Check out the ``master`` branch.

.. prompt:: bash host:~$, auto

    host:~$ git checkout master

In case you don't have a master branch yet, create one:

.. prompt:: bash host:~$, auto

    host:~$ git checkout -b master

You must store workflow files in the ``.github/workflows/`` directory of your repository.

.. prompt:: bash host:~$, auto

    host:~$ mkdir -p .github/workflows/

Finally, create the file ``mirror.yml`` and make sure you update the ``<FACTORY-NAME>`` with your Factory Name.

> .. note::
> 
>     Notice that this example is updating the FoundriesFactory repository ``containers.git``. 
>     The same approach could be used to update ``lmp-manifest.git`` and ``meta-subscriber-overrides.git``.

Reviewer 2: suggestion to move this
```suggestion
```

.. prompt:: bash host:~$, auto

    host:~$ gedit .github/workflows/mirror.yml

**.github/workflows/mirror.yml**:

.. prompt:: text

      name: Mirroring
      
      on: [push]
      
      jobs:
        to_foundries:
          runs-on: ubuntu-18.04
          steps:
            - uses: actions/checkout@v2
              with:
                fetch-depth: 0
            - uses: foundriesio/mirror-action@master
              with:
                REMOTE: "https://source.foundries.io/factories/<FACTORY-NAME>/containers.git"
                GIT_ACCESS_TOKEN: ${{ secrets.GIT_ACCESS_TOKEN }}
                PUSH_ALL_REFS: "false"

.. warning::

    ``PUSH_ALL_REFS`` is **false**. If **true** it synchronizes all branches.

.. warning::

    Make sure you make a backup of any content in your FoundriesFactory repository 
    that you want to preserve as this action can completely replace all branches.

> 

Reviewer 2: ```suggestion

.. note::

    Notice that this example is updating the FoundriesFactory repository ``containers.git``. 
    The same approach could be used to update ``lmp-manifest.git`` and ``meta-subscriber-overrides.git``.

```

Reviewer: The same approach can be used

> Add the changed files, commit and push:

Reviewer 2: is this done in the github repo?
```suggestion
Add the changed files, commit and push to your GitHub repository:
```

      .. prompt:: bash host:~$, auto

          host:~$ git add .github/workflows/mirror.yml
          host:~$ git commit -m "Adding Mirror Action"
          host:~$ git push

Github Action
-------------

> Once the ``mirror.yml`` is in place, everything you change to the configured 

Reviewer: When I configure a mirror on my github repo, does it mean I can push my commits to github and get it reflected to the FF repo? Or it it the other way around? 
I understand that I should be pushing to the FF repo, and github will only mirror what is in FF repo to the github repo, is that right?

Author: It is all done on github, I will update it to make it more clear.

> Once the ``mirror.yml`` is in place, everything you change to the configured 
> branch will start an Action on GitHub_ to mirror your repository.

Reviewer 2: ```suggestion
Once the ``mirror.yml`` is in place, every change to the configured 
branch will start an Action on GitHub_ to mirror your repository.
```

> You can find the GitHub_ Action by clicking on :guilabel:`Action`:

Reviewer 2: ```suggestion
You can find the GitHub_ Action by clicking on :guilabel:`Actions`:
```

.. figure:: /_static/userguide/mirror-action/mirror-action-github-action.png
   :width: 900
   :align: center

   GitHub Action

On that page, you can find the list of Actions as well as inspect each one.

.. figure:: /_static/userguide/mirror-action/mirror-action-github-action-list.png
   :width: 900
   :align: center

   GitHub Action list

> Compare your repositories from FoundriesFactory and GitHub_.

Reviewer 2: Changing this so we have an "ending/conclusion" feeling on this topic.
```suggestion
Your repositories from FoundriesFactory and GitHub_ should look the same.
```
